File: java/Statement.java
/******************************************************************************
 * Warren's Abstract Machine  -  Implementation by Stefan Buettcher
 *
 * developed:   December 2001 until February 2002
 *
 * Statement.java contains the class Statement, representing a single line of
 * WAM code, e.g. "true: proceed".
 ******************************************************************************/

import java.util.*;

// Statement class implements WAM code statements
// a statement looks like this:
// [label:] operator operand1 [operand2 [operand3]]
// label, op2 and op3 may be omitted. label is needed for jumps (calls)
public class Statement {
  public static final int opAllocate        =  1;
  public static final int opBigger          =  2;
  public static final int opCall            =  3;
  public static final int opCreateVariable  =  4;
  public static final int opCut             =  5;
  public static final int opDeallocate      =  6;
  public static final int opGetConstant     =  7;
  public static final int opGetValue        =  8;
  public static final int opGetVariable     =  9;
  public static final int opHalt            = 10;
  public static final int opIs              = 11;
  public static final int opGetLevel        = 12;
  public static final int opNoOp            = 13;
  public static final int opProceed         = 14;
  public static final int opPutConstant     = 15;
  public static final int opPutValue        = 16;
  public static final int opPutVariable     = 17;
  public static final int opRetryMeElse     = 18;
  public static final int opSmaller         = 19;
  public static final int opTrustMe         = 20;
  public static final int opTryMeElse       = 21;
  public static final int opUnifyList       = 22;
  public static final int opUnifyStruc      = 23;
  public static final int opUnequal         = 24;
  public static final int opUnifyVariable   = 25;
  public static final int opBiggerEq        = 27;
  public static final int opSmallerEq       = 28;
  public static final int opNotCall         = 29;

  public static final int callWrite         = -10;
  public static final int callWriteLn       = -11;
  public static final int callNewLine       = -12;
  public static final int callConsult       = -13;
  public static final int callReconsult     = -14;
  public static final int callLoad          = -15;
  public static final int callAssert        = -16;
  public static final int callRetractOne    = -17;
  public static final int callRetractAll    = -18;
  public static final int callIsInteger     = -19;
  public static final int callIsAtom        = -20;
  public static final int callIsBound       = -21;
  public static final int callReadLn        = -22;
  public static final int callCall          = -23;

  private String label;      // the label (used for jumping hin und her)
  private String function;   // the operator
  private Vector args;       // the operands vector
  public int operator;       // same as function, but as integer (performance!)
  public int jump;           // for faster jumping: if operand = call, then lookup target line number at startup
  public String arg1, arg2, arg3;   // for faster argument access from WAM

  // creates a new statement with one operand/argument
  public Statement(String aLabel, String aFunction, String anArgument) {
    label = aLabel.trim();
    function = aFunction.trim();
    args = new Vector();
    args.addElement(anArgument);
    args.addElement("");
    args.addElement("");
    doCommonStuff();
  } // end of Statement.Statement(String, String, String)

  // creates a new statement with one operand/argument
  public Statement(String aLabel, String aFunction, String arg1, String arg2) {
    label = aLabel.trim();
    function = aFunction.trim();
    args = new Vector();
    args.addElement(arg1);
    args.addElement(arg2);
    args.addElement("");
    doCommonStuff();
  } // end of Statement.Statement(String, String, String, String)

  // creates a new statement with more than two ops
  public Statement(String aLabel, String aFunction, String arg1, String arg2, String arg3) {
    int i;
    label = aLabel.trim();
    function = aFunction.trim();
    args = new Vector();
    args.addElement(arg1);
    args.addElement(arg2);
    arg3 = arg3;
    while ((i = arg3.indexOf(" ")) > 0) {
      args.addElement(arg3.substring(0, i));
      arg3 = arg3.substring(i + 1);
    }
    if (arg3.length() > 0)
      args.addElement(arg3);
    if (args.size() < 2)
      args.addElement("");
    doCommonStuff();
  } // end of Statement.Statement(String, String, String, String, String)

  private void doCommonStuff() {
    jump = -1;
    operator = functionToInt(function);
    arg1 = (String)args.elementAt(0);
    arg2 = (String)args.elementAt(1);
    arg3 = (String)args.elementAt(2);
  } // end of Statement.doCommonStuff()

  public int functionToInt(String function) {
    if (function.compareTo("allocate") == 0)          return opAllocate;
    if (function.compareTo("bigger") == 0)            return opBigger;
    if (function.compareTo("biggereq") == 0)          return opBiggerEq;
    if (function.compareTo("call") == 0)              return opCall;
    if (function.compareTo("not_call") == 0)          return opNotCall;
    if (function.compareTo("create_variable") == 0)   return opCreateVariable;
    if (function.compareTo("cut") == 0)               return opCut;
    if (function.compareTo("deallocate") == 0)        return opDeallocate;
    if (function.compareTo("get_constant") == 0)      return opGetConstant;
    if (function.compareTo("get_value") == 0)         return opGetValue;
    if (function.compareTo("get_variable") == 0)      return opGetVariable;
    if (function.compareTo("get_level") == 0)         return opGetLevel;
    if (function.compareTo("halt") == 0)              return opHalt;
    if (function.compareTo("is") == 0)                return opIs;
    if (function.compareTo("proceed") == 0)           return opProceed;
    if (function.compareTo("put_constant") == 0)      return opPutConstant;
    if (function.compareTo("put_value") == 0)         return opPutValue;
    if (function.compareTo("put_variable") == 0)      return opPutVariable;
    if (function.compareTo("retry_me_else") == 0)     return opRetryMeElse;
    if (function.compareTo("trust_me") == 0)          return opTrustMe;
    if (function.compareTo("try_me_else") == 0)       return opTryMeElse;
    if (function.compareTo("unequal") == 0)           return opUnequal;
    if (function.compareTo("unify_list") == 0)        return opUnifyList;
    if (function.compareTo("unify_struc") == 0)       return opUnifyStruc;
    if (function.compareTo("unify_variable") == 0)    return opUnifyVariable;
    if (function.compareTo("smaller") == 0)           return opSmaller;
    if (function.compareTo("smallereq") == 0)         return opSmallerEq;
    if ((function.compareTo("nop") == 0) || (function.compareTo("noop") == 0)) return opNoOp;
    return -1;
  } // end of Statement.functionToInt()

  // returns the label name of the statement
  public String getLabel() {
    return label;
  } // end of Statement.getLabel()

  // sets the label name to newLabel
  public void setLabel(String newLabel) {
    label = newLabel;
  } // end of Statement.setLabel(String)

  // returns the operator string, e.g. "get_variable"
  public String getFunction() {
    return function;
  } // end of Statement.getFunction()

  public void setFunction(String newFunction) {
    function = newFunction;
    operator = functionToInt(function);
  } // end of Statement.setFunction

  // returns the operand strings vector
  public Vector getArgs() {
    return args;
  } // end of Statement.getArgs()

  // for code dumping: print the statement: "label: operator op1 op2"
  public String toString() {
    if (label.compareTo(";") == 0)
      return "; " + function;
    String result;
    if (label.length() > 0) {
      result = label + ": ";
      for (int i = 1; i <= 12 - label.length(); i++)
        result += " ";
    }
    else
      result = "              ";
    result += function;
    for (int i = 0; i < args.size(); i++) {
      String a = (String)args.elementAt(i);
      if (a.indexOf(' ') < 0)
        result += " " + a;
      else
        result += " '" + a + "'";
    }
    if (jump >= 0)
      result += " (" + jump + ")";
    return result;
  } // end of Statement.toString()

  public String toString2() {
    String result = function;
    for (int i = 0; i < args.size(); i++)
      result += " " + args.elementAt(i);
    return result.trim();
  } // end of Statement.toString2()

  // where do you have to go today?
  public void setJump(int anAddress) {
    jump = anAddress;
  } // end of Statement.setJump(int)

  // where do you want to go today?
  public int getJump() {
    return jump;
  } // end of Statement.getJump()

} // end of class Statement

File: Statement.php
<?php

/* * ****************************************************************************
 * Warren's Abstract Machine  -  Implementation by Stefan Buettcher
 *
 * developed:   December 2001 until February 2002
 *
 * Statement.java contains the class Statement, representing a single line of
 * WAM code, e.g. "true: proceed".
 * **************************************************************************** */

// Statement class implements WAM code statements
// a statement looks like this:
// [label:] operator operand1 [operand2 [operand3]]
// label, op2 and op3 may be omitted. label is needed for jumps (calls)
class Statement {
    const opAllocate = 1;
    const opBigger = 2;
    const opCall = 3;
    const opCreateVariable = 4;
    const opCut = 5;
    const opDeallocate = 6;
    const opGetConstant = 7;
    const opGetValue = 8;
    const opGetVariable = 9;
    const opHalt = 10;
    const opIs = 11;
    const opGetLevel = 12;
    const opNoOp = 13;
    const opProceed = 14;
    const opPutConstant = 15;
    const opPutValue = 16;
    const opPutVariable = 17;
    const opRetryMeElse = 18;
    const opSmaller = 19;
    const opTrustMe = 20;
    const opTryMeElse = 21;
    const opUnifyList = 22;
    const opUnifyStruc = 23;
    const opUnequal = 24;
    const opUnifyVariable = 25;
    const opBiggerEq = 27;
    const opSmallerEq = 28;
    const opNotCall = 29;

    const callWrite = -10;
    const callWriteLn = -11;
    const callNewLine = -12;
    const callConsult = -13;
    const callReconsult = -14;
    const callLoad = -15;
    const callAssert = -16;
    const callRetractOne = -17;
    const callRetractAll = -18;
    const callIsInteger = -19;
    const callIsAtom = -20;
    const callIsBound = -21;
    const callReadLn = -22;
    const callCall = -23;

    private $label;      // the label (used for jumping hin und her)
    private $fonction;   // the operator
    private $args = array();       // the operands vector
    public $operator;       // same as function, but as integer (performance!)
    public $jump;           // for faster jumping: if operand = call, then lookup target line number at startup
    public $arg1, $arg2, $arg3;   // for faster argument access from WAM

    // creates a new statement with one operand/argument

    public function __construct($aLabel, $aFunction, $anArgument, $arg2 = '', $arg3 = '') {
        $this->label = trim($aLabel);
        $this->fonction = trim($aFunction);
        $this->args = array();
        $this->args[] = $anArgument;
        $this->args[] = $arg2;
        if (strlen($arg3)) {
            $rest = explode(' ', $args3);
            foreach ($rest as $item)
                $this->args[] = $item;
        } else
            $this->args[] = "";
        $this->doCommonStuff();
    }

// end of Statement.Statement(String, String, String)

    private function doCommonStuff() {
        $this->jump = -1;
        $this->operator = $this->functionToInt($this->fonction);
        $this->arg1 = $this->args[0];
        $this->arg2 = $this->args[1];
        $this->arg3 = $this->args[2];
    }

// end of Statement.doCommonStuff()

    public function functionToInt($fonction) {
        $mappingOp = array(
            "allocate" => self::opAllocate,
            "bigger" => self::opBigger,
            "biggereq" => self::opBiggerEq,
            "call" => self::opCall,
            "not_call" => self::opNotCall,
            "create_variable" => self::opCreateVariable,
            "cut" => self::opCut,
            "deallocate" => self::opDeallocate,
            "get_constant" => self::opGetConstant,
            "get_value" => self::opGetValue,
            "get_variable" => self::opGetVariable,
            "get_level" => self::opGetLevel,
            "halt" => self::opHalt,
            "is" => self::opIs,
            "proceed" => self::opProceed,
            "put_constant" => self::opPutConstant,
            "put_value" => self::opPutValue,
            "put_variable" => self::opPutVariable,
            "retry_me_else" => self::opRetryMeElse,
            "trust_me" => self::opTrustMe,
            "try_me_else" => self::opTryMeElse,
            "unequal" => self::opUnequal,
            "unify_list" => self::opUnifyList,
            "unify_struc" => self::opUnifyStruc,
            "unify_variable" => self::opUnifyVariable,
            "smaller" => self::opSmaller,
            "smallereq" => self::opSmallerEq,
            "nop" => self::opNoOp,
            "noop" => self::opNoOp
        );

        return array_key_exists($fonction, $mappingOp) ? $mappingOp[$fonction] : -1;
    }

// end of Statement.functionToInt()
    // returns the label name of the statement
    public function getLabel() {
        return $this->label;
    }

// end of Statement.getLabel()
    // sets the label name to newLabel
    public function setLabel($newLabel) {
        $this->label = $newLabel;
    }

// end of Statement.setLabel(String)
    // returns the operator string, e.g. "get_variable"
    public function getFunction() {
        return $this->fonction;
    }

// end of Statement.getFunction()

    public function setFunction($newFunction) {
        $this->fonction = $newFunction;
        $this->operator = $this->functionToInt($this->fonction);
    }

// end of Statement.setFunction
    // returns the operand strings vector
    public function getArgs() {
        return $this->args;
    }

// end of Statement.getArgs()
    // for code dumping: print the statement: "label: operator op1 op2"
    public function __toString() {
        if ($this->label == ";")
            return "; " . $this->fonction;
        // String result;
        if (strlen($this->label) > 0) {
            $result = $this->label . ": ";
            $result = str_pad($result, 14, ' ', STR_PAD_RIGHT);
        }
        else
            $result = str_repeat(" ", 14);
        $result .= $this->fonction;
        foreach ($this->args as $a) {
            if (false !== strpos($a, ' '))
                $result .= " " . $a;
            else
                $result .= " '$a'";
        }
        if ($this->jump >= 0)
            $result .= " (" . $this->jump . ")";
        return $result;
    }

// end of Statement.toString()

    public function toString2() {
        $result = $this->fonction;
        foreach ($this->args as $item)
            $result .= " " . $item;
        return trim($result);
    }

// end of Statement.toString2()
    // where do you have to go today?
    public function setJump($anAddress) {
        $this->jump = $anAddress;
    }

// end of Statement.setJump(int)
    // where do you want to go today?
    public function getJump() {
        return $this->jump;
    }

// end of Statement.getJump()
}

// end of class Statement
